Feat: Resource detail component - General metadata

0xLodz opened the pull request:
This PR implements https://github.com/okp4/dataverse-portal/issues/69 & https://github.com/okp4/dataverse-portal/issues/87.

Local testings can be performed on dataverse item pages: `/dataverse/[dataverseItem]/[id]`
Example for the screenshot below: `http://127.0.0.1:5173/dataverse/dataset/1`

![image](https://user-images.githubusercontent.com/8344874/228540343-161895de-d964-4172-bab5-6720870e5758.png)

## Fallback to `-` when Temporal Coverage metadata value is not an ISOstring.
![image](https://user-images.githubusercontent.com/8344874/229565827-339c4e47-a4fa-4d20-8e6e-8546197af68c.png)

fredericvilcot commented:
@0xLodz 
Please, rebase your branch from `origin/main` before opening a PR, because this PR embeds commits from another branches and does not target `main`.
Thx

bot-anik reviewed (commented):
![](https://img.shields.io/badge/Accessibility-82-orange?style=flat-square) ![](https://img.shields.io/badge/Best%20Practices-100-green?style=flat-square) ![](https://img.shields.io/badge/Performance-87-orange?style=flat-square) ![](https://img.shields.io/badge/Progressive%20Web%20App-25-red?style=flat-square) ![](https://img.shields.io/badge/SEO-100-green?style=flat-square) 
| Device | URL |
|--|--|
| desktop | http://localhost:4173/ |

![](https://img.shields.io/badge/Accessibility-82-orange?style=flat-square) ![](https://img.shields.io/badge/Best%20Practices-100-green?style=flat-square) ![](https://img.shields.io/badge/Performance-97-green?style=flat-square) ![](https://img.shields.io/badge/Progressive%20Web%20App-33-red?style=flat-square) ![](https://img.shields.io/badge/SEO-100-green?style=flat-square) 
| Device | URL |
|--|--|
| mobile | http://localhost:4173/ |

Not what you expected? Are your scores flaky? **GitHub runners could be the cause.**
[Try running on Foo instead](https://www.foo.software/docs/lighthouse-check-github-action/examples#running-on-foo-and-saving-results)
<!-- generated by lighthouse-check -->
<!-- COMMENT_ID["desktop:http://localhost:4173/","mobile:http://localhost:4173/"]COMMENT_ID -->

fredericvilcot commented:
@0xLodz 
I am a little bit confused about this PR, you say it implements #69, but how can we test the implementation requested in the feature as your component is imported nowhere?
If this PR aims to implement the feature, it should match the acceptance criteria asked from the product teal, that is obviously not the case.
Thanks .

ccamel commented on src/component/resource-details/resourceDetails.tsx:
```diff
@@ -0,0 +1,20 @@
+import type { FC } from 'react'
+
+type ResourceDetailsProps = {
+  icon: React.ReactNode
```
https://dev.to/ridhoanshory/jsx-element-react-node-react-element-idi

0xLodz commented:
> @0xLodz I am a little bit confused about this PR, you say it implements #69, but how can we test the implementation requested in the feature as your component is imported nowhere? If this PR aims to implement the feature, it should match the acceptance criteria asked from the product team, that is obviously not the case. Thanks .

@fredericvilcot sorry for the confusion. This feature has the PR #138 as dependency, that's why I initially set [feat/add-dataverse-items-dynamic-routes](https://github.com/okp4/dataverse-portal/tree/feat/add-dataverse-items-dynamic-routes) that implements #138 as the target branch, in order to provide only commits related to this feature. After reading you [previous message](https://github.com/okp4/dataverse-portal/pull/141#issuecomment-1487479202), I changed the target branch and removed the commit that provided a way to test the feature. In this case I don't see any other solution than changing the target branch from `main` to [feat/add-dataverse-items-dynamic-routes](https://github.com/okp4/dataverse-portal/tree/feat/add-dataverse-items-dynamic-routes). What would you recommend?
Thanks.

fredericvilcot commented:
@0xLodz 
To keep a clean and consistant history, we need to wait for the merge of the blocking pull requests.
Please switch this one to draft meanwhile.
Thank you.

ErikssonJoakim commented on src/page/dataverse/component/generalMetadata/generalMetadata.tsx:
```diff
@@ -0,0 +1,57 @@
+
+export type GeneralMetadata = {
+  iconName: IconName
+  title: GeneralMetadataTitle
```
We cannot specify the different titles in the component because the title value depends on the caller. For example it makes no sense for the data space page to have the format title because a data space does not have a format meta data

ErikssonJoakim reviewed (changes requested):
Great work 😃 There are some things to consider but it's a good start!

0xLodz commented on src/page/dataverse/dataset/dataset.tsx:
```diff
@@ -1,9 +1,50 @@
+type MatchedDatasetProps = {
+  label: string
+}
+const MatchedDataset: FC<MatchedDatasetProps> = ({ label }) => (
```
I do agree that this abstraction is unnecessary, it add some noise in the file. We discussed about it with @lolottetheclash. I did that by providing more readability at the `match` function level. I am ok to remove it. What do you think @lolottetheclash ?

lolottetheclash replied:
@0xLodz I think that this decision is personal and that we do not all share the same point of view, I will say that it is not decisive, up to you ;)

ccamel commented on src/page/dataverse/component/generalMetadata/generalMetadata.tsx:
```diff
@@ -0,0 +1,57 @@
+  | 'temporalCoverage'
+  | 'category'
+
+export type GeneralMetadata = {
```
There are three distinct GeneralMetadata models: one for Dataspace, one for Service, and one for Dataspace. While it may seem possible to merge them into one model, doing so would result in inaccuracies in the meaning and nature of the conveyed information. Therefore, it is advisable to rely on the concepts of the ontology for consistency as much as possible.

ccamel replied:
Definitively the second option. But, please take a step back, and consider the element to be displayed as a metadata property (and not a general metadata title), which describes a property (what you call the title), its value (what you call description), to which you associate an icon, eventually:

```ts
type MetadataProperty = {
  property: string
  value: string
  iconName?: IconName
}
```

You may not know this yet, but ontology properties, like the format property (`core:hasFormat`) are also described as such, and therefore have a label property (`rdfs:label`), internationalized, which specifies the label to use when representing the property. This is why you can't use: `t(`generalMetadata.${title}`)`. So it makes sense that the property is a string, since the value conveys the exact representation of the property, for the current local language.

Now, another issue arises with the values. You are assuming that it is a string, but that is not necessarily the case. For instance, the temporal coverage is a date range, and the dates must be represented according to the locale. And this is the responsibility of the component, imho... This is just fantastic 😅

ccamel commented on src/ui/view/dataverse/component/generalMetadata/generalMetadata.tsx:
```diff
@@ -0,0 +1,54 @@
+  iconName?: IconName
+}
+
+const GeneralMetadataItem: FC<MetadataProperty> = memo(
```
Can you explain the purpose of `memo` here. Thanks :pray:

0xLodz replied:
In that case it avoids unnecessary re-renders of the `GeneralMetadataItem` by memoizing it's output according to its props, If it's parent re-render for any reason, it won't be re-rendered. However this component is quite small, it might be worse to apply this optimisation mostly on components that are expensive to render.

ccamel commented on src/util/isoDateTime/isoDateTime.ts:
```diff
@@ -0,0 +1,15 @@
+const getYearFromISODateTime = (ISODateTime: string, locale: string): string => {
+  const date = new Date(ISODateTime)
+  const formatter = new Intl.DateTimeFormat(locale, { year: 'numeric' })
+  return formatter.format(date)
+}
```
In the function, I would delegate the responsibility of determining whether a string is an ISO DateTime to the `Date` constructor. Moreover, as it stands, if you pass a string that does not comply with the format, the function throws an exception.

fredericvilcot replied:
Yes, let's delegate the responsibility to the Date constructor and treat returned value (or Error) before calling `convertToYearIfISODateTime` method: it will prevent from returning a non formatted input.

0xLodz replied:
What do you think about a validation that use both the regex and the `Date` constructor to determine if a string is an ISODateTime? Because for example `new Date(1)` is a date.

> Moreover, as it stands, if you pass a string that does not comply with the format, the function throws an exception.
You are right, that the reason why I didn't exported the function. That's probably too implicit and needs error handling if we want to use it directly.

ccamel replied:
Yes. My previous comment was inaccurate. You should rely on the `Date.parse()` function to do what you want without needing a regex. Just check that the number returned is not `NaN`, and it will work as expected.

fredericvilcot commented on src/ui/page/dataverse/dataset/dataset.tsx:
```diff
@@ -15,7 +45,14 @@ const Dataset = (): JSX.Element => {
     () => <p>Dataset not found</p>,
-    (dataset: DataverseItemDetails) => <p>{dataset.label}</p>
+    (dataset: DataverseItemDetails) => {
+      return (
```
Please keep ES6 syntax

fredericvilcot commented on src/ui/page/dataverse/dataspace/dataspace.tsx:
```diff
@@ -15,7 +25,14 @@ const Dataspace = (): JSX.Element => {
     () => <p>Dataspace not found</p>,
-    (dataspace: DataverseItemDetails) => <p>{dataspace.label}</p>
+    (dataspace: DataverseItemDetails) => {
+      return (
```
Please keep ES6 syntax

fredericvilcot commented on src/ui/page/dataverse/service/service.tsx:
```diff
@@ -15,7 +25,14 @@ const Service = (): JSX.Element => {
     () => <p>Service not found</p>,
-    (service: DataverseItemDetails) => <p>{service.label}</p>
+    (service: DataverseItemDetails) => {
+      return (
```
Please keep ES6 syntax

fredericvilcot commented on src/util/isoDateTime/isoDateTime.spec.ts:
```diff
@@ -0,0 +1,18 @@
+    ['2023-01-01T00:00:00.000-01:00', true],
+    ['002023-01-01T00:00:00.000+01:00', false]
+  ])('validates input %s as %s', (input: string, expected: boolean) => {
+    expect(isISODateTime(input)).toBe(expected)
```
IMHO you should only test the `convertToYearIfISODateTime` method that will test all other methods in a traversal way and ensures you 100% of coverage.

0xLodz replied:
Thanks for pointing this out. That makes sense to get a full coverage.

fredericvilcot reviewed (changes requested):
Great job, I have some remarks to be discussed, thx

fredericvilcot commented on src/util/isoDateTime/isoDateTime.ts:
```diff
@@ -0,0 +1,15 @@
+  return isoDateTimeRegex.test(input) && new Date(input).toString() !== 'Invalid Date'
+}
+
+export const convertToLocalizedYearIfISODateTime = (input: string, locale: string): string => {
+  return isISODateTime(input) ? getLocalizedYear(new Date(input), locale) : input
+}
```
You should not return the `input`: what about if I give `foo` as the `input` parameter of the `convertToLocalizedYearIfISODateTime` method? => the view will display `foo` as a temporal coverage, that is a wrong interpretation.
In a functional way, we would have used a `Either` task to capture both values (result, error) and then handle it in the view.
For now, I'll suggest either to throw an error or return a `Invalid date` that will be directly handle as a fallback in the view. (content must be validated with the product team)

fredericvilcot replied:
And please, keep ES6 syntax 😉

ccamel replied:
@0xLodz The point is: _never trust the input_. The function `convertToLocalizedYearIfISODateTime` should function consistently in _any_ scenario, with the following expected behavior:

- It should return the year of the input date if it can be successfully parsed
- If it cannot be parsed, an error should be returned. The approach to return the error is at your discretion: _functional style_ (the prefered one) using an [Either<>](https://gcanti.github.io/fp-ts/modules/Either.ts.html) value, or _imperative style_ by throwing an error, as suggested by @fredericvilcot

lolottetheclash reviewed (approved):
👌

ccamel reviewed (approved):
Let's go! Adjustments will be made when the data model is available
